### PR TITLE
REF: avoid getattr pattern in libgroupby rank functions

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @WillAyd do we need to do anything special w/r/t the comment on L1220 re #19560?

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -419,13 +419,21 @@ def get_func(fname):
...
-                if f is not None:
-                    return f
+                if f2 is not None:
+                    return f2
+
+            if hasattr(f, "__signatures__"):
+                # inspect what fused types are implemented
+                if dtype_str == "object" and "object" not in f.__signatures__:
+                    # return None so we get a NotImplementedError below
+                    #  instead of a TypeError at runtime
```
Comment (gfyoung) [inline pandas/core/groupby/ops.py]:
```suggestion
                    # instead of a TypeError at runtime
```

File: pandas/_libs/groupby.pyx [hunk 2]
```diff
@@ -1213,14 +1201,6 @@ def group_rank(float64_t[:, :] out,
...
 
 
-group_rank_float64 = group_rank["float64_t"]
-group_rank_float32 = group_rank["float32_t"]
-group_rank_int64 = group_rank["int64_t"]
-group_rank_uint64 = group_rank["uint64_t"]
-# Note: we do not have a group_rank_object because that would require a
-#  not-nogil implementation, see GH#19560
-
-
```
Comment (gfyoung) [inline pandas/_libs/groupby.pyx]: Are these just unused?

**Comment by WillAyd** [comment]
> cc @WillAyd do we need to do anything special w/r/t the comment on L1220 re #19560?

Is this changing behavior when trying to rank objects? I think on master raises an error

```python
>>> pd.DataFrame([[1]], dtype=object).groupby([1]).rank()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/groupby.py", line 2144, in rank
    axis=axis,
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/groupby.py", line 867, in _cython_transform
    result, _ = self.grouper.transform(obj.values, how, **kwargs)
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/ops.py", line 621, in transform
    return self._cython_operation("transform", values, how, axis, **kwargs)
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/ops.py", line 583, in _cython_operation
    result, values, labels, func, is_numeric, is_datetimelike, **kwargs
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/ops.py", line 658, in _transform
    transform_func(result, values, comp_ids, ngroups, is_datetimelike, **kwargs)
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/ops.py", line 441, in wrapper
    return f(afunc, *args, **kwargs)
  File "/Users/williamayd/clones/pandas/pandas/core/groupby/ops.py", line 393, in <lambda>
    kwargs.get("na_option", "keep"),
TypeError: 'NoneType' object is not callable
```

Though not sure how well tested that is. Ideally would be explicitly disallowed but looks like that PR is still open, so might want to be careful when changing this

File: pandas/_libs/groupby.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby.pyx]: If you look at get_cython_function in groupby.ops, you'll see we do a thing with checking for `libgroupby func_name` and then `libgroupby.func_name_dtype`.  This pattern was put in place back before we started using fused types for libgroupby, and now cython effectively choose the correct func_name_dtype for us when we just use func_name

**Comment by jbrockmendel** [comment]
> Is this changing behavior when trying to rank objects? I think on master raises an error

The same example now raises `pandas.core.base.DataError: No numeric types to aggregate`

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**